### PR TITLE
feat!: network customization and tooling improvements

### PR DESCRIPTION
**Pull request by antazoey**
### What I did

* `NetworkBoundCommand` renamed to `ConnectedProviderCommand` and no longer requires explicitly setting the network option (base wasn't removed and warns a DeprecatedWarning with instructions how to migrate and it has the legacy behavior).
* Both `@network_option` and `ConnectedProviderCommand` allow key-args in your callback for `ecosystem`, `network,` and `provider` for each instance respectively. You don't need to use any of them either (yay, can delete all the `_ = network  # needed for NetworkBoundCommand` lines).
* Move another class to `ape_ethereum` called `EthereumNodeProvider`.. It is exactly what was in ape-geth, however ape-geth still sorta re-exports because `EthereumNodeProvider` is the default class for custom networks and such and just sounds better and in some cases may be a better base class than `Web3Provider` alone. So on the node, `adhoc` neworks are rebranded as `custom` networks so itll be smoother to add the config-based ones with names and chaind IDs and such.
* Adds some other Deprecated methods back to less the damager of 0.7 on scripts and emit Deprecated warnings and TODOs to delete on 0.8 which **is going to happen**.

fixes: #1572 
Fixes: APE-1239
fixes: #1401 
Fixes: APE-881
Fixes: APE-486

### How I did it

when you use `@network_option` now, it analyzes the fn  you give it via `inspect` and then it checks out `sys.argv` to parse the value, since at option time it does not have access to it. It then turns the callback fn to a partial with the fields filled out.

`ConnectedProviderCommand` (f.k.a NetworkBoundCommand) changes were much easier to implement because it has access to code responsible for pass kwargs into your command callback.. so at that time, it is able to put the right ones in, also via using `inspect` to see what the callback is even requesting.

But because `NetworkBoundCommand` no longer needs `network_option` and most likely never will use it together because it is pointless unless you are being super custom with it, it will never have the partially-filled callback.. Thus it utiliizes is own way of passing in the instances..

it was a little more complex than i anticipated, mostly because Options are parsed so early on. If this only happened in `ConnectedProviderCommand`, it'd be easier, but I got it working nonetheless.

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by antazoey** [comment]
Now, we won't need unused `network` values in our CLIs like:

```
_ = network  # Used for NeworkBoundCommand
```

And also if we do need it, it is actually useful.

**Comment by antazoey** [comment]
after this PR, I want to all config-based adhoc network customization so it doesnt have to say "adhoc" and you can tell it the name and chain Id and whatever else

File: docs/userguides/clis.md [hunk 1]
```diff
@@ -63,28 +63,60 @@ def cli(cli_ctx):
...
+--network ethereum
+--network ethereum:sepolia
+--network ethereum:mainnet:alchemy
+--network ::foundry
+```
+
+To use default values automatically, omit sections of the choice, but leave the semi-colons for parsing.
+For example, `::test` means use the default ecosystem and network and the `test` provider.
+
+Use `ecosystem`, `network`, and `provider` argument names in your command implementation to access their corresponding class instances:
```
Comment (fubuloubu) [inline docs/userguides/clis.md:79]: Nice

**Review by fubuloubu — APPROVED** [review]
General thoughts:

1. When this goes out, we need to educate folks how to upgrade their CLIs or it may cause problems
2. Conversely, we can have `NetworkBoundCommand` just raise an error demonstrating how to upgrade across the 0.7.x series

And a question:
1. I can use `@network_option` without `ConnectedProviderCommand` and it won't connect?

**Comment by antazoey** [comment]
> 1. I can use `@network_option` without `ConnectedProviderCommand` and it won't connect?

Correct, using the option does not connect, but that isn't a changing behavior, it already works that way.

And yes, good point, I can some deprecated methods back in and do the deprecated warning thing.

**Comment by antazoey** [comment]
@fubuloubu A good example of using `@network_option` without being connected is the `ape cache init` methods, which don't need to connect but sill needs the network value given for creating the DB. I made sure that still works.

Also :

- [x] Put back `NetworkBoundCommand` with deprecation warning and legacy behavior.

**Comment by antazoey** [comment]
- [ ] fix a wrong-instance issue

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**